File: nodejs/get_labels.js
const fs = require('fs');
const readline = require('readline');

// Create readline interface
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let MIRO_ORG_ID;
let API_TOKEN;

// Function to ask a question
function askQuestion(question, validator, callback) {
    rl.question(question, (answer) => {
        if (validator(answer)) {
            callback(answer);
        } else {
            console.log('Invalid input. Please try again.');
            askQuestion(question, validator, callback); // Re-ask the question
        }
    });
}

// Array of questions with corresponding validators
const questions = [
    {
      question: 'Enter your Miro Organization ID: ',
      validator: (answer) => !isNaN(parseFloat(answer)) && isFinite(answer)
    },
    {
        question: 'Enter your Miro REST API Token: ',
        validator: (answer) => typeof answer === 'string'
    }
    // Add more questions with validators as needed
];

// Function to ask multiple questions recursively
async function askQuestions(index) {

    if (index >= questions.length) {
        // End of questions
        console.log('Thank you for answering the questions!');
        await init(MIRO_ORG_ID, API_TOKEN);
        rl.close();
        return;
    }
  
    const { question, validator } = questions[index];
    askQuestion(question, validator, (answer) => {
        if (question === 'Enter your Miro Organization ID: ') {
            MIRO_ORG_ID = answer.toString();
        }
        else if (question === 'Enter your Miro REST API Token: ') {
            API_TOKEN = answer.toString();;
        }
        askQuestions(index + 1); // Ask the next question
    });
}

// Start asking questions
askQuestions(0);

async function callAPI(url, options) {
    async function manageErrors(response) {
        if(!response.ok){
            const parsedResponse = await response.json();
            const responseError = {
                status: response.status,
                statusText: response.statusText,
                requestUrl: response.url,
                errorDetails: parsedResponse
            };
            throw(responseError);
        }
        return response;
    }

    const response = await fetch(url, options)
    .then(manageErrors)
    .then((res) => {
        if (res.ok) {
            const rateLimitRemaining = res.headers.get('X-RateLimit-Remaining');
            return res[res.status == 204 ? 'text' : 'json']().then((data) => ({ status: res.status, rate_limit_remaining: rateLimitRemaining, body: data }));
        }
    })
    .catch((error) => {
        console.error('Error:', error);
        return error;
    });
    return response;
}

function jsonToCsv(jsonData) {
    let csv = '';
    // Get the headers
    let headers = Object.keys(jsonData[Object.keys(jsonData)[0]]);
    csv += headers.join(',') + '\n';
    // Add the data
    Object.keys(jsonData).forEach(function(row) {
        let data = headers.map(header => JSON.stringify(jsonData[row][header])).join(','); // Add JSON.stringify statement
        csv += data + '\n';
    });
    return csv;
}

async function init(orgId, token) {
    const apiUrl = `https://api.miro.com/v2/orgs/${orgId}/data-classification-settings`;
    const reqHeaders = {
        'cache-control': 'no-cache, no-store',
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ' + token
    };
    const reqGetOptions = {
        method: 'GET',
        headers: reqHeaders,
        body: null
    };
    try {
        const getClassificationLabels = await callAPI(apiUrl, reqGetOptions);
        console.log('============== CLASSIFICATION LABELS - BEGIN ===================');
        console.log(JSON.stringify(getClassificationLabels.body, null, 2));
        console.log('=============== CLASSIFICATION LABELS - END ===================');
        if (getClassificationLabels.status === 200) {
            const getClassificationLabelsArray = [];
            for(let i=0; i < getClassificationLabels.body.labels.length; i++) {
                var label = {
                    label_id: getClassificationLabels.body.labels[i].id,
                    label_name: getClassificationLabels.body.labels[i].name,
                    is_default: getClassificationLabels.body.labels[i].default,
                    description: getClassificationLabels.body.labels[i].description ? getClassificationLabels.body.labels[i].description : '',
                    order_number: getClassificationLabels.body.labels[i].orderNumber,
                    type: getClassificationLabels.body.labels[i].type
                };
                getClassificationLabelsArray.push(label);
            }
            const directory = 'board_classification_labels';
            if (!fs.existsSync(directory)) {
                fs.mkdirSync(directory);
            }
            let content;
            let filePath;
            content = jsonToCsv(getClassificationLabelsArray);
            filePath = 'board_classification_labels/classification_labels.csv';
            fs.writeFileSync(filePath, content);

            content = JSON.stringify(getClassificationLabelsArray, null, 2);
            filePath = 'board_classification_labels/classification_labels.json';
            fs.writeFileSync(filePath, content);
            
            console.log('# Next steps:\n# 1. Review the classification labels from the list above (or open the "classification_labels.csv" file within the folder "board_classification_labels" in the directory where this script lives)\n# 2. Identify the label you want to use to classify the unclassified boards (you will be asked for the ID of the desired label on step 3)\n# 3. Run: node classification.js');
            console.log('===========================================');
        }
    }
    catch(error) {
        console.log(error);
    }
}

File: python/get_labels.py
import asyncio
import json
import aiohttp
import os

# Define global variables to store answers
MIRO_ORG_ID = 0
API_TOKEN = ''

def json_to_csv(json_data):
    # Convert JSON keys to a list if they are not already in that format
    if isinstance(json_data, dict):
        json_data = [json_data[key] for key in json_data]

    # Get the headers from the first element of the JSON data
    headers = list(json_data[0].keys())
    csv_data = ','.join(headers) + '\n'
    
    # Add the data
    for row in json_data:
        data = ','.join(json.dumps(row[header]) for header in headers)
        csv_data += data + '\n'
    
    return csv_data

# def json_to_csv(json_data):
#     if json_data:
#         csv = ''
#         # Get the headers
#         headers = list(json_data[0].keys())
#         csv += ','.join(headers) + '\n'
        
#         # Helper function to escape CSV special characters
#         def escape_csv(value):
#             if isinstance(value, (int, float)):
#                 value = str(value)
#             if isinstance(value, str):
#                 if '"' in value:
#                     value = value.replace('"', '""')
#                 # if ',' in value or '"' in value or '\n' in value:
#             value = f'"{value}"'
#             return value
        
#         # Add the data
#         for row in json_data:
#             data = ','.join(escape_csv(json_data[row][header]) for header in headers)
#             csv += data + '\n'
        
#         return csv

async def call_api(url, options):
    async def manage_errors(response):
        if not response.status == 200:
            parsed_response = await response.json()
            response_error = {
                'status': response.status,
                'status_text': response.reason,
                'request_url': str(response.url),
                'error_details': parsed_response
            }
            raise Exception(response_error)
        return response

    async with aiohttp.ClientSession() as session:
        async with session.request(url=url, **options) as response:
            try:
                response = await manage_errors(response)

                rate_limit_remaining = response.headers.get('X-RateLimit-Remaining')
                if response.status == 204:
                    body = await response.text()
                else:
                    body = await response.json()

                return {
                    'status': response.status,
                    'rate_limit_remaining': rate_limit_remaining,
                    'body': body
                }
            except Exception as e:
                return {
                    'status': response.status,
                    'status_text': response.reason,
                    'request_url': str(response.url),
                    'error_details': str(e)
                }

async def get_labels(session, org_id):
    global MIRO_ORG_ID
    global API_TOKEN

    req_headers = {
        'cache-control': 'no-cache, no-store',
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {API_TOKEN}'
    }

    req_get_options = {
        'method': 'GET',
        'headers': req_headers,
        'data': None
    }

    url = f'https://api.miro.com/v2/orgs/{org_id}/data-classification-settings'
    print(f'============== CLASSIFICATION LABELS - BEGIN ===================')
    list_labels = await call_api(url, req_get_options)
    print(json.dumps(list_labels, indent=4))
    print(f'============== CLASSIFICATION LABELS - END ===================')
    
    if list_labels['status'] == 200:
        get_classification_labels_array = []
        for item in list_labels['body']['labels']:
            label = {
                'label_id': item['id'],
                'label_name': item['name'],
                'is_default': item['default'],
                'description': item['description'] if 'description' in item else '',
                'order_number': item['orderNumber'],
                'type': item['type']
            }
            get_classification_labels_array.append(label)
        
        directory = 'board_classification_labels'
        if not os.path.exists(directory):
            os.makedirs(directory)

        with open('board_classification_labels/classification_labels.json', 'w') as file:
            file.write(str(get_classification_labels_array))

        content = json_to_csv(get_classification_labels_array)
        with open('board_classification_labels/classification_labels.csv', 'w') as file:
            file.write(str(content))
        
        print(f'# Next steps:\n# 1. Review the classification labels from the list above (or open the "classification_labels.csv" file within the folder "board_classification_labels" in the directory where this script lives)\n# 2. Identify the label you want to use to classify the unclassified boards (you will be asked for the ID of the desired label on step 3)\n# 3. Run: python3 get_labels.py')
        print('===========================================')
        return True

    else:
        print('====== see errors array below ======')
        result = {
            'organization_id': org_id,
            'response_error': json.dumps(list_labels),
            'responde_status': list_labels['status']
        }
        print(json.dumps(result, indent=4))
        return False

async def init():
    async with aiohttp.ClientSession() as session:
        await get_labels(session, MIRO_ORG_ID)
    return True

def ask_number_question(question):
    while True:
        answer = input(question).strip()
        if answer.isdigit():
            return int(answer)
        else:
            print("Invalid answer. Please enter a valid number.")

def ask_any_input_question(question):
    return input(question).strip()

def ask_questions():
    global MIRO_ORG_ID
    global API_TOKEN

    MIRO_ORG_ID = ask_number_question('Enter your Miro Organization ID: ')
    API_TOKEN = ask_any_input_question('Enter your Miro REST API Token: ')

def after_questions_answered():
    print("Thank you for answering the questions!")
    asyncio.run(init())

ask_questions()
after_questions_answered()
